### ISSUE TITLE
Fuzz for memory disclosure vulnerabilities

### ISSUE THREAD

**Issue by Shnatsel** [kornelski/lodepng-rust#32]
The current fuzzing harness allows detecting out of bounds memory access as well as panics and any kind of crashes. It will not detect memory disclosure vulnerabilities, which can be devastating in image libraries ([example](http://seclists.org/fulldisclosure/2013/Nov/83)).

I have dissected such a vulnerability in DEFLATE decoder used in `png` crate in [my recent Reddit post](https://www.reddit.com/r/rust/comments/8zpp5f/auditing_popular_crates_how_a_oneline_unsafe_has/) and outlined the ways it could be discovered via fuzzing. Sadly I do not have the time to set it up such a fuzzing harness for lodepng myself.

**Comment by kornelski** [comment]
What kind of setup you'd recommend for that?

**Comment by Shnatsel** [comment]
There are two options:

1. Use memory sanitizer (not to be confused with address sanitizer). However, any C code included in the binary also needs to be instrumented, otherwise msan will produce false positives. This will probably involve rebuilding libc crate with memory sanitizer included, although I'm not sure as I've never managed to do that myself.

2. Create a fuzz harness that decodes the same image twice and make sure the output is the same. This only works as long as the buffers that the decoding code uses are not reused. AFAIK Rust's default jemalloc may reuse memory allocations; switching to the system allocator (as stabilized in Rust 1.28) may or may not fix this problem; I'm not sufficiently familiar with the low-level details of it to be sure.

**Comment by Shnatsel** [comment]
AFL's [libdislocator](https://github.com/mirrorer/afl/tree/master/libdislocator) is a really good candidate for malloc replacement. It's designed to expose memory bugs.

**Comment by Shnatsel** [comment]
I've experimented with this a bit using a [trivial testcase](https://gist.github.com/Shnatsel/0c024a51b64c6e0b6c6e66f991904816). You cannot detect uninitialized memory use by comparing output of a function called twice in the same process with any allocator - be it system, jemalloc or libdislocator. Only with system allocator does the output vary between runs, but only if the program is restarted. With jemalloc and libdislocate it does not vary at all!

Fortunately, libdislocate code is trivial, so I've easily [patched](https://gist.github.com/Shnatsel/d00ca9a1080676897f08b81e461a1abe) it to clobber every allocated buffer with a new value. This allows reliably detecting use of uninitialized memory simply by running image decoding twice and comparing the output.

**Comment by Shnatsel** [comment]
I've cleaned up that hacky patch on top of libdislocator. It now has [its own repository](https://github.com/Shnatsel/libdiffuzz) with a fancy name and without the memory leak inherent to the design of libdislocator.

I also made a differential fuzzing target for lodepng, which has discovered no issues after just over half a billion executions under AFL. I've cleaned that up too and put it up on github as an example, see https://github.com/Shnatsel/lodepng-afl-fuzz-differential
